shotover-proxy/src/message/mod.rs

```diff
@@ -426,45 +423,112 @@ impl Value {
-                }
-                ColType::Uuid => Value::Bytes(Bytes::copy_from_slice(actual_bytes)),
-                ColType::Varchar => Value::Strings(decode_varchar(actual_bytes).unwrap()),
-                ColType::Varint => unimplemented!("We dont have a varint type yet"),
-                ColType::Timeuuid => Value::Bytes(Bytes::copy_from_slice(actual_bytes)),
-                ColType::Inet => Value::Inet(decode_inet(actual_bytes).unwrap()),
-                ColType::Timestamp => Value::NULL,
-                ColType::Date => Value::NULL,
-                ColType::Time => Value::NULL,
-                ColType::Smallint => {
-                    Value::Integer(decode_smallint(actual_bytes).unwrap() as i64, IntSize::I16)
-                }
-                ColType::Tinyint => {
-                    Value::Integer(decode_tinyint(actual_bytes).unwrap() as i64, IntSize::I8)
-                }
-                // TODO: process collection types based on ColTypeOption
-                // (https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v4.spec#L569)
-                _ => Value::NULL,
+        let cassandra_type = Value::into_cassandra_type(&spec.col_type, data);
+        Value::create_element(cassandra_type)
+    }
+
+    fn into_cassandra_type(col_type: &ColTypeOption, data: &CBytes) -> CassandraType {
+        let wrapper = get_wrapper_fn(&col_type.id);
+        wrapper(data, col_type)
+    }
+
+    fn create_element(element: CassandraType) -> Value {
+        match element {
+            CassandraType::Ascii(a) => Value::Ascii(a),
+            CassandraType::Bigint(b) => Value::Integer(b, IntSize::I64),
+            CassandraType::Blob(b) => Value::Bytes(b.into_vec().into()),
+            CassandraType::Boolean(b) => Value::Boolean(b),
+            CassandraType::Counter(c) => Value::Counter(c),
+            CassandraType::Decimal(d) => {
+                let big_decimal = BigDecimal::new(d.unscaled, d.scale.into());
+                Value::Decimal(big_decimal)
             }
+            CassandraType::Double(d) => Value::Double(d.into()),
+            CassandraType::Float(f) => Value::Float(f.into()),
+            CassandraType::Int(c) => Value::Integer(c as i64, IntSize::I64),
+            CassandraType::Timestamp(t) => Value::Timestamp(t),
+            CassandraType::Uuid(u) => Value::Uuid(u),
+            CassandraType::Varchar(v) => Value::Varchar(v),
+            CassandraType::Varint(v) => Value::Varint(v),
+            CassandraType::Timeuuid(t) => Value::Timeuuid(t),
+            CassandraType::Inet(i) => Value::Inet(i),
+            CassandraType::Date(d) => Value::Date(d),
+            CassandraType::Time(d) => Value::Time(d),
+            CassandraType::Smallint(d) => Value::Integer(d.into(), IntSize::I16),
+            CassandraType::Tinyint(d) => Value::Integer(d.into(), IntSize::I8),
+            CassandraType::List(_) => Value::create_list(element),
```

Comment:
We could drastically simplify this by moving the logic from the functions directly into the match.

![image](https://user-images.githubusercontent.com/5120858/149064430-1aaa3bad-9a5a-4bbd-b448-a7a966195b6c.png)

This removes the need for the panic and the logic is small enough to comfortably fit in the match